File: app/src/main/java/net/tjado/usbgadget/GadgetViewModel.java
package net.tjado.usbgadget;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;

public class GadgetViewModel extends AndroidViewModel {

    private Application application;

    private MutableLiveData<List<GadgetObject>> gadgetData;
    private MutableLiveData<Boolean> rootGranted;

    GadgetShellApi gsa;
    GadgetAssetProfiles gap;

    String [] gadgetProfileList;

    public GadgetViewModel(@NonNull Application app) {
        super(app);
        application = app;

        gadgetData = new MutableLiveData<>();
        gadgetData.setValue(new ArrayList<>());

        rootGranted = new MutableLiveData<>();
        rootGranted.setValue(true);

        this.gsa = new GadgetShellApi();
        this.gap = new GadgetAssetProfiles(getApplication());
        gadgetProfileList = gap.getAllGadgetProfiles();
        updateGadgetData();
    }

    public MutableLiveData<Boolean> hasRootPermissions() {
        return rootGranted;
    }

    public MutableLiveData<List<GadgetObject>> getGadgets() {
        return gadgetData;
    }

    public void reloadGadgetData() {
        List<GadgetObject> data = gadgetData.getValue();
        gadgetData.setValue(data);
    }

    public void updateGadgetData() {
        this.gsa.updateGadgetData(gadgetData, rootGranted);
    }

    public void loadGadgetProfile(String gadget) {
        String profile = this.gap.getProfileGadget(gadget);

        this.gsa.exec(profile, response -> {
            // update gadget data if new gadget was loaded
            updateGadgetData();
        });
    }

    public void loadFunctionProfile(GadgetObject gadget, String function) {
        String profile = this.gap.getProfileFunction(function, gadget.getValue("gadget_path"));

        this.gsa.exec(profile, response -> {
            // update gadget data if new gadget was loaded
            updateGadgetData();
        });
    }

}

File: app/src/main/java/net/tjado/usbgadget/GadgetViewModel.kt
package net.tjado.usbgadget

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import java.util.ArrayList

class GadgetViewModel(application: Application) : AndroidViewModel(
    application
) {
    val gadgets: MutableLiveData<List<GadgetObject>>
    private val rootGranted: MutableLiveData<Boolean>

    private val shellApi: GadgetShellApi
    private val assets: GadgetAssetProfiles

    val gadgetProfileList: Array<String>

    init {
        gadgets = MutableLiveData()
        gadgets.value = ArrayList()

        rootGranted = MutableLiveData()
        rootGranted.value = true

        shellApi = GadgetShellApi()
        assets = GadgetAssetProfiles(getApplication())
        gadgetProfileList = assets.allGadgetProfiles
        updateGadgetData()
    }

    fun hasRootPermissions(): MutableLiveData<Boolean> {
        return rootGranted
    }

    fun reloadGadgetData() {
        val data = gadgets.value!!
        gadgets.value = data
    }

    fun updateGadgetData() {
        shellApi.updateGadgetData(gadgets, rootGranted)
    }

    fun loadGadgetProfile(gadget: String) {
        val profile = assets.getProfileGadget(gadget)
        if (profile != null) {
            shellApi.exec(profile) {
                // update gadget data if new gadget was loaded
                updateGadgetData()
            }
        }
    }

    fun loadFunctionProfile(gadget: GadgetObject, function: String) {
        val profile = assets.getProfileFunction(function, gadget.getValue("gadget_path"))
        if (profile != null) {
            shellApi.exec(profile) {
                // update gadget data if new gadget was loaded
                updateGadgetData()
            }
        }
    }
}